### PR TITLE
Refactor the whole project

### PR DESCRIPTION
**Pull request by Coeuvre**
The new texture packer has three components: `Importer`, `Packer` and `Exporter`.

`Importer` is used to import something to `Texture` while `Exporter` will export `Texture` to other things. The magic is anyone who implement `Texture` can be packed by `Packer`.

`Packer` has configurations to control his behavior such as:

* which algorithm
* allow rotation
* shape padding
* border padding
* inner padding
* trim mode
* shape outline (for debug)
* etc

Those configurations are inspired by [TexturePacker](https://www.codeandweb.com/texturepacker/) and still under development. I'll focus on `Skyline` algorithm first.

The solution of color conversion is inspired by @TyOverby in his recently PR #18.

### CODE REVIEW

**Comment by bvssvni** [comment]
@Coeuvre Nice to see you back!

**Comment by Coeuvre** [comment]
:smile:

**Comment by TyOverby** [comment]
Did you forget to remove buffer2d?

File: src/texture_packer.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+        TexturePacker {
+            textures: HashMap::new(),
+            frames: HashMap::new(),
+            packer: packer,
+            config: config,
+        }
+    }
+
+    pub fn pack(&mut self, key: String, texture: Box<Texture<Pixel = P> + 'a>) {
```
Comment (TyOverby) [inline src/texture_packer.rs]: This should probably be rewritten as 

```rust
pub fn pack<T: Texture<Pixel = P>>(&mut self, key: String, texture: &T)
```

File: src/exporter/image_exporter.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+use image::{
+    DynamicImage,
+    ImageBuffer,
+    Rgba,
+};
+
+use texture::Texture;
+
+#[derive(Copy)]
+pub struct ImageExporter<T>;
```
Comment (TyOverby) [inline src/exporter/image_exporter.rs]: Should this be a struct with no fields or just make `export` a public function?

Comment (TyOverby) [inline src/exporter/image_exporter.rs]: ImageExporter should be implemented by default for `DynamicImage`.

**Comment by TyOverby** [comment]
My main concern with the rewrite is that instead of generics, there is quite a bit of boxing (of textures and of packers).  In the case of TexturePacker::pack, you actually have to give ownership of the texture to the texture-packer, which I think could be replaced with references or a `std::borrow::Cow`.

File: src/packer/skyline_packer.rs [hunk 3]
```diff
@@ -155,39 +151,39 @@ impl<B: Buffer2d> SkylinePacker<B> {
...
+    type Pixel = P;
 
-        match self.find_skyline(width, height) {
-            Some((i, mut rect)) => {
-                if width == rect.w {
-                    self.buf.patch(rect.x, rect.y, buf);
-                } else {
-                    self.buf.patch_rotated(rect.x, rect.y, buf);
-                }
+    fn pack(&mut self, key: String, texture: &Texture<Pixel=P>) -> Option<Frame> {
```
Comment (TyOverby) [inline src/packer/skyline_packer.rs]: Why did you choose strings as keys?

Comment (Coeuvre) [inline src/packer/skyline_packer.rs]: Because we need some ways to retrieve the frame after packing texture. Do you have suggestions?

**Pull request merged**